tracing-subscriber/src/layer.rs

```diff
@@ -0,0 +1,255 @@
+use tracing_core::{
+    metadata::Metadata,
+    span,
+    subscriber::{Interest, Subscriber},
+    Event,
+};
+
+use crate::filter;
+use std::any::TypeId;
+
+pub trait Layer<S>: 'static {
+    /// Registers a new callsite with this layer, returning whether or not
+    /// the subscriber is interested in being notified about the callsite.
+    ///
+    /// This function is provided with the `Interest` returned by the wrapped
+    /// subscriber. The layer may then choose to return
+    fn register_callsite(&self, _metadata: &'static Metadata<'static>, prev: Interest) -> Interest {
+        prev
+    }
+    fn enabled(&self, _metadata: &Metadata, prev: bool) -> bool {
+        prev
+    }
+    fn new_span(&self, _attrs: &span::Attributes, _id: &span::Id) {}
+    fn record(&self, _span: &span::Id, _values: &span::Record) {}
+    fn record_follows_from(&self, _span: &span::Id, _follows: &span::Id) {}
+    fn event(&self, _event: &Event) {}
+    fn enter(&self, _id: &span::Id) {}
+    fn exit(&self, _id: &span::Id) {}
+    fn clone_span(&self, _id: &span::Id, _new: Option<&span::Id>) {}
```

Comment:
Hmm, something that's occurred to me is that we probably _don't_ actually want to have `clone_span` and `drop_span` hooks on `Layer` --- instead, we probably want to only ever forward them to the actual `Subscriber`, and have a method on `layer` that gets called on _close_ --- i.e., if a refcount hit 0. The problem is that _this_ means we need a way for a `Subscriber` to signal a close with a return value, which `drop_span` doesn't currently have.

We could probably add a method like `try_close(&self, &span::Id) -> bool` that's like drop_span but returns if the ref count is zero, and I think it's actually possible to add this to `Subscriber` non-breakingly (with a default impl that never returns `true`), but this feels kind of awkward IMO.

@LucioFranco @jonhoo @davidbarsky what do y'all think? Is it okay for `Layer`s to have to manage their own independent ref-counts?

Reply:
@hawkw yeah, I believe that should wor, currently the only work I do within drop_span that isnt ref counting is on close https://github.com/timberio/vector/blob/master/lib/trace-metrics/src/lib.rs#L210. So for my use case that is totally appropriate, I think.

Reply:
The downside here is that adding something to the `Subscriber` API in a backwards compatible way, the potential API gets awkward.

Reply:
What if it was possible to add an API in a breaking fashion? I feel that `tracing` ought make as many breaking changes as it possibly can, _while_ it still can.

Reply:
@davidbarsky I'm definitely not opposed to making breaking changes when necessary, and I agree that it's better to make them now than wait until it's really painful later. 

However, I'm not convinced this _needs_ a breaking change. I took a crack at exploring this API in #148. I think the proposed non-breaking change is somewhat less ugly than I'd initially anticipated, and doesn't really make the API surface that much more confusing, provided we add good documentation. Let me know what you think?

Reply:
> However, I'm not convinced this needs a breaking change. I took a crack at exploring this API in #149. I think the proposed non-breaking change is somewhat less ugly than I'd initially anticipated, and doesn't really make the API surface that much more confusing, provided we add good documentation. Let me know what you think?

As discussed, I like that API! Given the API change is rather minimal, I think its appropriate to deprecate `drop_span` in tracing 0.1 and drop it wholesale in tracing 0.2.

Reply:
I'm fine with that, though I think the deprecation should be done in a separate change, since it would necessitate fixing a bunch of dependent crates.

Reply:
> I'm fine with that, though I think the deprecation should be done in a separate change, since it would necessitate fixing a bunch of dependent crates.

I'm in agreement. There's no rush to do that now.